controllers/bpfman-agent/application-program.go

```diff
@@ -0,0 +1,322 @@
+package bpfmanagent
+
+import (
+	"context"
+	"fmt"
+
+	bpfmaniov1alpha1 "github.com/bpfman/bpfman-operator/apis/v1alpha1"
+	"github.com/bpfman/bpfman-operator/internal"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+//+kubebuilder:rbac:groups=bpfman.io,resources=bpfapplications,verbs=get;list;watch
+
+type BpfApplicationReconciler struct {
+	ReconcilerCommon
+	currentApp *bpfmaniov1alpha1.BpfApplication
+	ourNode    *v1.Node
+}
+
+// TODO: As implemented, the BpfApplicationReconciler doesn't need to implement
+// the bpfmanReconciler interface.  We should think about what's needed and what
+// isn't.
+
+func (r *BpfApplicationReconciler) getRecType() string {
+	return internal.ApplicationString
+}
+
+func (r *BpfApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Initialize node and current program
+	r.currentApp = &bpfmaniov1alpha1.BpfApplication{}
+	r.ourNode = &v1.Node{}
+	r.Logger = ctrl.Log.WithName("application")
+	r.appOwner = &bpfmaniov1alpha1.BpfApplication{}
+	r.finalizer = internal.BpfApplicationControllerFinalizer
+	r.recType = internal.ApplicationString
+
+	ctxLogger := log.FromContext(ctx)
+	ctxLogger.Info("Reconcile Application: Enter", "ReconcileKey", req)
+
+	// Lookup K8s node object for this bpfman-agent This should always succeed
+	if err := r.Get(ctx, types.NamespacedName{Namespace: v1.NamespaceAll, Name: r.NodeName}, r.ourNode); err != nil {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed getting bpfman-agent node %s : %v",
+			req.NamespacedName, err)
+	}
+
+	appPrograms := &bpfmaniov1alpha1.BpfApplicationList{}
+
+	opts := []client.ListOption{}
+
+	if err := r.List(ctx, appPrograms, opts...); err != nil {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed getting BpfApplicationPrograms for full reconcile %s : %v",
+			req.NamespacedName, err)
+	}
+
+	if len(appPrograms.Items) == 0 {
+		r.Logger.Info("BpfApplicationController found no application Programs")
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	var res ctrl.Result
+	var err error
+	var complete bool
+
+	for i, a := range appPrograms.Items {
+		for j, p := range a.Spec.Programs {
+			switch p.Type {
+			case bpfmaniov1alpha1.ProgTypeFentry:
+				fentryProgram := bpfmaniov1alpha1.FentryProgram{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: a.Name + "-fentry",
```

Comment:
Not sure yet if we need this name to be deterministic or not but you can use `generateName: a.Name` if not https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values

Then we never need to worry about hardcoded postfixes

Reply:
switched to generateName

Reply:
This name is used to generate the BpfProgram object names, and those need to be unique and deterministic because we need to use the same name every time we run the reconciler.So, if we randomly generated a name for each program in the application object, we'd need to save it, and I'm not sure where we would do that unless we create a new CRD.  

This scheme has a limitation in that we can only support one program of each type for a given attach point (otherwise, we'd get duplicate names).  Mohamed thought this was acceptable because it doesn't make sense for an application to do that. 

Just thinking -- if this limitation isn't acceptable, would it be possible to have an optional name field for each program in the AppProgram CRD and let the operator controller generate it globally and save it in the AppProgram CRD if the user doesn't provide one?

Reply:
yeah it doesn't make sense to have more than one kprobe for example attached to the same exact attach point in an application however we can have two kprobes one attached to `tcp_rcv_established` and another attached to `psample_sample_packet` just as an example

Reply:
Sounds good I'm fine with leaving it, makes everything more readable